### PR TITLE
Integrated in-app onboarding tours

### PR DESCRIPTION
**Pull request by arrenv**
## Description

The product tour for Colony creation is an interactive, step-by-step guide highlighting the main components and functionalities of the Colony interface while walking them through the process of setting up their colony. It will activate immediately after a user creates their first colony, via certain actions, or manually via the new "Guided tours" user menu item, offering a mix of visual and textual cues to guide them through different aspects of the platform.

## Testing

I have focused on the foundations and put together one complete tour "Using the app'. Future tours will be added in separate PRs, as there has been troubles with the opening and open status of the action panel.

**Manual tour trigger - ColonyContext**
1. Click on the user menu in the top left.
2. Click on the "Guided tours" menu item.
3. Click on the "Using the app" item.
4. Go through the tour.

**Manual tour trigger - Not ColonyContext**
1. Click on the user menu in the top left.
2. Click on the "Guided tours" menu item.
3. There should be an empty state if no tours are available. 

Note: Currently tours are also not available on mobile, so you get the same empty state.

## Diffs

**New stuff** ✨

* New TourContext has been added.
* New data attribute is being used on tags.
* Tour steps have been added.

image

image

image

## TODO

- [x] Complete full first tour.
- [ ] Add manually triggered additional tours:
    - [ ] ~~Get started~~ (Add in separate PR)
    - [ ] ~~Adding funds~~ (Add in separate PR)
    - [ ] ~~Making payments~~ (Add in separate PR)
    - [x] Using the app
    - [ ] ~~User dashboard~~ (Add in separate PR)
- [ ] ~~Add auto-triggers for relevant tours~~ (Add in separate PR)
    - [ ] ~~Get started should auto trigger and be conditional for colony creators or just new users.~~ (Add in separate PR)

### CODE REVIEW

File: src/components/common/Extensions/UserNavigation/partials/UserSubmenu/partials/Tours/Tours.tsx
Comment (iamsamgibbs) [inline src/components/common/Extensions/UserNavigation/partials/UserSubmenu/partials/Tours/Tours.tsx:122]: `if (isTablet)`

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Great job on this Arren! Looks like it was a big undertaking, especially with the rebases involved! It is all working very nicely.

Tested in both light and dark mode, and on really big window size all working without issue.
 
Screenshot 2024-11-12 at 12 02 08

Only a couple of small things I think it is worth resolving before getting it merged. The tour doesn't work on tablet, but still shows in the menu:

Screenshot 2024-11-12 at 12 05 16

Mobile is fine:

Screenshot 2024-11-12 at 12 05 11

The arrow colour on dark mode should match the rest of the pop up.

Screenshot 2024-11-12 at 11 59 32

And the final step of the tour could say "Done" rather than "Next".

Screenshot 2024-11-12 at 12 01 30

Nice work on this, looks like a really good foundation for adding more tours.

**Comment by iamsamgibbs** [comment]
Oh! Forgot to mention, I had to `rm -rf node_modules`, remove the package-lock.json and run `npm install` to get the frontend to build.

Screenshot 2024-11-12 at 11 25 54

**Comment by jakubcolony** [comment]
Same issue here @iamsamgibbs, perhaps a rebase on latest master would fix it?

File: src/context/TourContext/TourContextProvider.tsx
Comment (jakubcolony) [inline src/context/TourContext/TourContextProvider.tsx:38]: Do we need this mapping? It just re-adds properties that already exist because of the spread, apart from `data.title`. At the same time `data.title` doesn't seem to be used anywhere in our codebase (but I'm not sure if it's maybe needed by react-joyride).

File: src/context/ColonyCreateModalContext/ColonyCreateModalContextProvider.tsx
Comment (jakubcolony) [inline src/context/ColonyCreateModalContext/ColonyCreateModalContextProvider.tsx:21]: Is there a use case for closing the modals like so yet? I couldn't trigger a tour while any of the modals was open. I'm wondering if it would be better to only add this once needed.

Also since that similar block of code is repeated across a few components, a shared hook sounds like a good idea. We should also use an enum for the event name.

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Nice work @arrenv, the colony tour is looking great:

image
image
image
image
image
image

However, I seem to be unable to open the action form when clicking on any of the cards:

https://github.com/user-attachments/assets/46d3b29e-24d1-40c4-8e5d-49d239d71e25

I also left a few in-code comments, would love to hear your thoughts

File: src/components/common/Tours/enums.ts
Comment (arrenv) [reply src/components/common/Tours/enums.ts:1]: Agreed, there was more done, but scaled back for the initial tour, cleaning up now.

File: src/components/common/Tours/createAction/messages.tsx
Comment (arrenv) [reply src/components/common/Tours/createAction/messages.tsx:1]: Yep, agreed.

File: src/context/ActionSidebarContext/ActionSidebarContext.ts
Comment (arrenv) [reply src/context/ActionSidebarContext/ActionSidebarContext.ts:10]: Good pickup, I removed the usage for now, so will undo.

File: src/context/ColonyCreateModalContext/ColonyCreateModalContextProvider.tsx
Comment (arrenv) [reply src/context/ColonyCreateModalContext/ColonyCreateModalContextProvider.tsx:21]: Which modals did you test? I implemented it due to the issue occurring when opening the tour while the token locking modals were open, I have just tested again and is needed. I agree though, this would be better as a shared hook and I will use an enum.

File: src/context/TourContext/TourContextProvider.tsx
Comment (arrenv) [reply src/context/TourContext/TourContextProvider.tsx:38]: Not anymore, good pickup!

File: src/context/ColonyCreateModalContext/ColonyCreateModalContextProvider.tsx
Comment (jakubcolony) [reply src/context/ColonyCreateModalContext/ColonyCreateModalContextProvider.tsx]: You're right, I must've missed that one

**Review by jakubcolony — APPROVED** [review]
Thank you for making the changes, the whole setup looks really good now.

Re-tested the tour and it still works as expected:
test-tour


I can confirm I'm no longer seeing the npm issue.

And the action cards are clickable again:
test-action-cards

**Review by iamsamgibbs — APPROVED** [review]
Thanks for the changes, all looking great! Also had no issues with npm this time.

I still think the tip of the arrow should match the rest of the pop up, but it's not enough of an issue to prevent merging here.

Screenshot 2024-12-04 at 10 50 27

Everything else is great, no longer showing on tablet.

Screenshot 2024-12-04 at 10 51 55

"Finish tour" button on the final step

Screenshot 2024-12-04 at 10 50 42

No tours show when outside of the colony context

Screenshot 2024-12-04 at 10 52 23

Great job, lets get it merged!

**Review by rdig — APPROVED** [review]
All good here tested and works as advertised. No tours for user context, and with a guided tour for the colony context. 

Very nicely done! :100: 


https://github.com/user-attachments/assets/556e2c92-6284-4792-bd03-de1794f3bd30

The only thing different that I would have done, is put all the tour logic behind a feature flag, otherwise, if we ever want to take it down quickly _(ie: wrong info or something)_ we can just disable it from there, rather than scrambling for a new deployment

Another thing I noticed is that the shadow from the tooltip will render above the arrow, making it look like a weird line

Screenshot from 2024-12-05 11-39-12
Screenshot from 2024-12-05 11-39-19
Screenshot from 2024-12-05 11-39-33
Screenshot from 2024-12-05 11-39-38

I'm personally good with it going in like this, so I leave it up to you if you want to fix it or not

**Pull request merged**
